packages/cli/src/models/local/UnpackController.ts

```diff
@@ -0,0 +1,72 @@
+import path from 'path';
+import request from 'request-promise-native';
+import fs from 'fs-extra';
+
+import { Logger } from 'zos-lib';
+
+import stdout from '../../utils/stdout';
+import patch from '../../utils/patch';
+import cp from '../../utils/cp';
+
+const ora = patch('ora');
+const simpleGit = patch('simple-git/promise');
+
+interface KitConfig {
+  ignore: string[],
+  hooks: Object,
+}
+
+export default class UnpackController {
+  public async unpack(url: string, workingDirPath: string = '', config: KitConfig): Promise<void | never> {
+    if (!url) throw Error('A url must be provided.');
+    if (!config) throw Error('A config must be provided.');
+
+    const { exec } = cp;
+    const { readdir, remove, pathExists } = fs;
+
+    // because zos always spawns '.zos.lock' file
+    if ((await readdir(workingDirPath)).length > 1) throw Error('The directory must be empty');
+
+    try {
+      let spinner = ora(`Downloading kit from ${url}`).start();
+      const git = simpleGit(workingDirPath);
+      await git.init();
+      await git.addRemote('origin', url);
+      await git.pull('origin', 'stable');
+      spinner.succeed();
+
+      spinner = ora(`Unpacking kit`).start();
+      // always delete .git folder 
+      config.ignore.push(`.git`);
+      // delete all files/folders from ignore
+      const paths = config.ignore.map(file => path.join(workingDirPath, file));
+      const exists = await Promise.all(paths.map(path => pathExists(path)));
+      await Promise.all(exists.map((ext, i)=> ext ? remove(paths[i]) : null ).filter(Boolean));
+      
+      // run kit commands like `npm install`
+      await exec(config.hooks['post-unpack']);
+      spinner.succeed();
+
+      stdout(`Done. Amazing!`);
+      stdout(`Please, continue at ${url}`);
+    } catch(e) {
+      // TODO: remove all files from directory on fail except .zos.lock
+      e.message = `Failed to download and unpack kit from ${url}. Details: ${e.message}`;
```

Comment:
It is actually used on ZOS to some extent like:
````
error.message = `${contractAlias} deployment failed with error: ${error.message}`;
throw error;
````
I am actually more concerned about code which spawns new error killing old stracktrace, like this
````
} catch(error) {
    throw new Error(error.message || 'Error while trying to verify contract');
  }
````
I think I'll start `Coding Guide` document rather soon so we all can collaborate and define guidelines for us and future developers to work on project.

Reply:
> I think I'll start Coding Guide document rather soon so we all can collaborate and define guidelines for us and future developers to work on project.

Sounds good. FWIW, I'd actually start by adopting a guide from an existing project, rather than draft one from scratch.

Reply:
Yes, I was thinking the same? Do you have project in mind?